solidity/random-beacon/test/RandomBeacon.Relay.test.ts

```diff
@@ -850,10 +766,569 @@ describe("RandomBeacon - Relay", () => {
+                  .withArgs(membersIDs.slice(0, 31), expectedUntil)
+              })
+
+              it("should pay notifier reward from heartbeat notifier rewards pool", async () => {
+                const expectedReward = ineligibleOperatorNotifierReward.mul(
+                  validFailedMembersIndices.length
+                )
+
+                const currentNotifierBalance = await testToken.balanceOf(
+                  notifier.address
+                )
+                expect(
+                  currentNotifierBalance.sub(initialNotifierBalance)
+                ).to.be.equal(expectedReward)
+
+                const currentHeartbeatNotifierRewardsPoolBalance =
+                  await randomBeacon.heartbeatNotifierRewardsPool()
+                expect(
+                  initialHeartbeatNotifierRewardsPoolBalance.sub(
+                    currentHeartbeatNotifierRewardsPoolBalance
+                  )
+                ).to.be.equal(expectedReward)
+              })
+            })
+
+            context("when one of the signatures is incorrect", () => {
+              const assertInvalidSignature = async (invalidSignature) => {
+                // The 32 signers sign correct parameters. Invalid signature
+                // is expected to be provided by signer 33.
+                const { signatures, signingMembersIndices } =
+                  await signHeartbeatFailureClaim(
+                    members,
+                    0,
+                    group.groupPubKey,
+                    validFailedMembersIndices,
+                    groupThreshold - 1
+                  )
+
+                await expect(
+                  randomBeacon.notifyFailedHeartbeat({
+                    groupId,
+                    failedMembersIndices: validFailedMembersIndices,
+                    // Slice removes `0x` prefix from wrong signature.
+                    signatures: signatures + invalidSignature.slice(2),
+                    signingMembersIndices: [...signingMembersIndices, 33],
+                  })
+                ).to.be.revertedWith("Invalid signature")
+              }
+
+              context(
+                "when one of the signatures signed the wrong nonce",
+                () => {
+                  it("should revert", async () => {
+                    // Signer 33 signs wrong nonce.
+                    const invalidSignature = (
+                      await signHeartbeatFailureClaim(
+                        [members[32]],
+                        1,
+                        group.groupPubKey,
+                        validFailedMembersIndices,
+                        1
+                      )
+                    ).signatures
+
+                    await assertInvalidSignature(invalidSignature)
+                  })
+                }
+              )
+
+              context(
+                "when one of the signatures signed the wrong group public key",
```

Comment:
Can we have test cases where:
1. `failedMembersIndices` contains just one member with index somewhere in the middle of group member indices.
2. `failedMembersIndices` contains multiple indices but they are not subsequent, e.g. [2,5,7,23,56].
3. `signingMembersIndices` contains multiple indices but they are not subsequent.

Reply:
Here they are: https://github.com/keep-network/keep-core/pull/2736/commits/5f4c04dac691263693fe054dd892bc648e8245e7